File: solutions/python/min-heap/03_peek/min_heap.py
class MinHeap:
    def __init__(self) -> None:
        self._items = []

    def insert(self, value) -> None:
        self._items.append(value)
        self._bubble_up(len(self._items) - 1)

    def peek_min(self):
        if not self._items:
            return None
        return self._items[0]

    def _bubble_up(self, index: int) -> None:
        current = index
        while current > 0:
            parent = (current - 1) // 2
            if self._items[parent] <= self._items[current]:
                break
            # Swap parent and current
            temp = self._items[current]
            self._items[current] = self._items[parent]
            self._items[parent] = temp
            current = parent

File: solutions/java/min-heap/step03_peek/MinHeap.java
import java.util.ArrayList;
import java.util.List;

public class MinHeap {
    private final List<Integer> items;

    public MinHeap() {
        this.items = new ArrayList<>();
    }

    public void insert(int value) {
        items.add(value);
        bubbleUp(items.size() - 1);
    }

    public Integer peekMin() {
        if (items.isEmpty()) {
            return null;
        }
        return items.get(0);
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }

    private void bubbleUp(int index) {
        int current = index;
        while (current > 0) {
            int parent = (current - 1) / 2;
            if (items.get(parent) <= items.get(current)) {
                break;
            }
            swap(parent, current);
            current = parent;
        }
    }

    private void swap(int i, int j) {
        int temp = items.get(i);
        items.set(i, items.get(j));
        items.set(j, temp);
    }
}
